packages/gatsby/src/redux/run-sift.js

```diff
@@ -103,37 +103,95 @@ function handleMany(siftArgs, nodes) {
  * A fast index is created if one doesn't exist yet so cold call is slower.
  * The empty result value is null if firstOnly is false, or else an empty array.
  *
- * @param {Array<string>} chain Note: `$eq` is assumed to be the leaf prop here
- * @param {boolean | number | string} targetValue chain.a.b.$eq === targetValue
+ * @param {Array<DbQuery>} filters Resolved. (Should be checked by caller to exist)
  * @param {Array<string>} nodeTypeNames
- * @param {Map<string, Map<string | number | boolean, Node>>} typedKeyValueIndexes
- * @returns {Array<Node> | undefined}
+ * @param {Map<string, Map<string | number | boolean, Set<IGatsbyNode>>>} typedKeyValueIndexes
+ * @returns {Array<IGatsbyNode> | undefined}
  */
-const runFlatFilterWithoutSift = (
-  chain,
-  targetValue,
+const runFlatFiltersWithoutSift = (
+  filters,
   nodeTypeNames,
   typedKeyValueIndexes
 ) => {
-  ensureIndexByTypedChain(chain, nodeTypeNames, typedKeyValueIndexes)
-
-  const nodesByKeyValue = getNodesByTypedChain(
-    chain,
-    targetValue,
+  const caches = getBucketsForFilters(
+    filters,
     nodeTypeNames,
     typedKeyValueIndexes
   )
 
-  // If we couldn't find the needle then maybe sift can, for example if the
-  // schema contained a proxy; `slug: String @proxy(from: "slugInternal")`
-  // There are also cases (and tests) where id exists with a different type
-  if (!nodesByKeyValue) {
+  if (!caches) {
+    // Let Sift take over as fallback
     return undefined
   }
 
-  // In all other cases this must be a non-empty Set because the indexing
-  // mechanism does not create a Set unless there's a Node for it
-  return [...nodesByKeyValue]
+  // Put smallest last (we'll pop it)
+  caches.sort((a, b) => b.length - a.length)
```

Comment:
Should get buckets for filters return pre-sorted caches? Maybe not.

Reply:
It's possible but not necessary for `eq`, which seems to be the most common case by far. We can tweak that later by benchmarking it.